lib/smart_proxy_dns_plugin_template/dns_plugin_template_main.rb

```diff
@@ -1,58 +1,38 @@
-require 'dns/dns'
+require 'dns_common/dns_common'
 
 module Proxy::Dns::PluginTemplate
   class Record < ::Proxy::Dns::Record
     include Proxy::Log
     include Proxy::Util
 
-    attr_reader :example_setting
+    attr_reader :example_setting, :optional_path, :required_setting, :required_path
 
-    def self.record(attrs = {})
-      new(attrs.merge(:example_setting => ::Proxy::Dns::PluginTemplate::Plugin.settings.example_setting))
+    def initialize
+      @required_setting = ::Proxy::Dns::PluginTemplate::Plugin.settings.required_setting # never nil
+      @example_setting = ::Proxy::Dns::PluginTemplate::Plugin.settings.example_setting # can be nil
+      @required_path = ::Proxy::Dns::PluginTemplate::Plugin.settings.required_path # file exists and is readable
+      @optional_path = ::Proxy::Dns::PluginTemplate::Plugin.settings.optional_path # nil, or file exists and is readable
+
+      # Common settings can be defined by the main plugin, it's ok to use them locally.
+      # Please note that providers must not rely on settings defined by other providers or plugins they are not related to.
+      super('localhost', ::Proxy::Dns::Plugin.settings.dns_ttl)
+    end
+
+    # Calls to these methods are guaranteed to have non-nil parameters
+    def create_a_record(fqdn, ip)
+      # FIXME: add a forward 'A' record with fqdn, ip
     end
 
-    def initialize options = {}
-      @example_setting = options[:example_setting]
-      raise "dns_plugin_template provider needs 'example_setting' option" unless example_setting
-      super(options)
+    def create_ptr_record(fqdn, ip)
+      # FIXME: add a reverse 'PTR' record with ip, fqdn
```

Comment:
Keep the example code to handle the collision? Same for the other create and remove.